### PR TITLE
Phase 10.6: MCP elicitation (real `onElicitation` translation)

### PR DESCRIPTION
**Pull request by TylerLeonhardt**
## What

Replaces the Phase-7 `onElicitation` cancel-stub for the Claude agent with a real translation. When an MCP server (the host's own in-process server, or a third-party server) calls `elicit/create`, the request now **surfaces to the user** as structured input instead of being silently auto-cancelled, and the user's answer is delivered back to the server as an `ElicitResult`.

Before:
```ts
onElicitation: async req => {
	logElicitation(req.message ?? '');
	return { action: 'cancel' };   // silently declined
},
```

## How

Elicitation routes through the **same structured user-input channel `AskUserQuestion` already uses** (`requestUserInput` → `ActionType.ChatInputRequested` → `respondToUserInputRequest`), **not** the `pending_confirmation` permission gate (design locked in `CONTEXT.md`).

- **`claudeElicitation.ts`** (new) — pure projections: JSON-Schema → `ChatInputQuestion[]` and workbench answers → `ElicitResult.content`. Each field is runtime-validated with the base-layer `vs/base/common/validation.ts` (drops malformed fields instead of throwing); the field type is *derived* from that validator and pinned to the MCP SDK's authoritative `PrimitiveSchemaDefinition` by a compile-time guard, so it can't drift from the spec.
- **`claudeElicitationBridge.ts`** (new) — `handleElicitation` looks up the session, parks on `requestUserInput`, observes the SDK per-request abort signal, and maps the response back to an `ElicitResult`. `decline` is reserved for an explicit user Decline; no-session / teardown / abort all map to `cancel`.
- Wire `onElicitation` through `IBuildOptionsInput` → `Options.onElicitation`, `IMaterializeContext`, and `ClaudeAgent._makeOnElicitation` — exactly mirroring `canUseTool`. The `logElicitation` stub param is removed.
- **`form`** mode (schema → questions) and **`url`** mode (question-less request carrying the URL) are both supported.

## Scope notes

- `SDKElicitationCompleteMessage` (out-of-band URL completion) is intentionally **out of scope** — the promise-await model already resolves URL prompts via the existing `ChatInputCompleted` path.
- Subagent-scoped routing (`parentToolCallId`) is out of scope; elicitation carries no `agentID`.
- No change to the `IAgent` public surface or the protocol `ChatInputRequest` shape.

Full step-by-step design + decisions: [`phase10.6-plan.md`](https://github.com/microsoft/vscode/blob/tyleonha/claude-phase-10.6-mcp-elicitation/src/vs/platform/agentHost/node/claude/phase10.6-plan.md).

## Testing

- **Unit** (`claudeElicitation.test.ts`, 11): every field kind + branch (text/number/integer/boolean/oneOf/enum±enumNames/array items.enum+anyOf/unknown+missing type), url mode, malformed-schema and every empty/broken-form → message-only degradation, per-field validation drop, accept/decline/cancel mapping, accept edge cases, bridge no-session → cancel.
- **Agent** (`claudeAgent.test.ts` Phase 10.6, 5): form accept round-trip, decline, abort-mid-park, url-mode round-trip, pre-aborted signal.
- **Integration** (`claudeAgent.integrationTest.ts` §5.3): real accept round-trip through the proxy-backed materialize → Options boundary.
- **Manual E2E**: verified live in the Agents window driving real Claude Opus against `@modelcontextprotocol/server-everything` — `trigger-elicitation-request` renders a 13-field form.

Local gate green: `typecheck-client`, `valid-layers-check`, eslint, full agentHost unit suite (2725), claude integration (5), claude protocol e2e (8).

🤖 Generated with assistance from Copilot.

### CODE REVIEW

File: src/vs/platform/agentHost/node/claude/claudeElicitation.ts [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+	if (response !== ChatInputResponseKind.Accept) {
+		return { action: 'cancel' };
+	}
+	const schema = request.mode === 'url' ? undefined : parseElicitationSchema(request.requestedSchema);
+	if (!schema) {
+		return { action: 'accept' };
+	}
+	const content: Record<string, ElicitationFieldValue> = {};
+	for (const [name] of schema.fields) {
+		const value = elicitationAnswerToValue(answers?.[name]);
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply src/vs/platform/agentHost/node/claude/claudeElicitation.ts]: Confirmed and fixed in 5db6c25. The consumer renders number/integer/boolean questions as text inputs (`agentHostSessionHandler.ts:2858-2874` has no dedicated cases) and `convertCarouselAnswers` emits `Text`, so `elicitationAnswerToValue` is now schema-aware — it takes the field and coerces the text answer to the declared type (`Math.trunc` for integers, `Number.isFinite` guards, boolean text→bool), dropping uncoercible values. New test: `elicitationResultFromAnswers coerces text answers to the field schema type`.

File: src/vs/platform/agentHost/node/claude/claudeElicitation.ts [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+				...base,
+				kind: field.type === 'integer' ? ChatInputQuestionKind.Integer : ChatInputQuestionKind.Number,
+				min: field.minimum,
+				max: field.maximum,
+				defaultValue: typeof field.default === 'number' ? field.default : undefined,
+			};
+		case 'array':
+			return {
+				...base,
+				kind: ChatInputQuestionKind.MultiSelect,
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply src/vs/platform/agentHost/node/claude/claudeElicitation.ts:216]: Fixed in 5db6c25 — the MultiSelect projection now sets `allowFreeformInput: false` (the consumer defaults an omitted value to `true` at line 2855). Snapshot tests updated.

File: src/vs/platform/agentHost/node/claude/claudeElicitation.ts [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+				max: field.maxItems,
+			};
+		case 'string':
+		default:
+			// Titled single-select (`oneOf`), enum/legacy single-select (`enum`,
+			// optionally `enumNames`), or a plain text field.
+			if (field.oneOf) {
+				return {
+					...base,
+					kind: ChatInputQuestionKind.SingleSelect,
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply src/vs/platform/agentHost/node/claude/claudeElicitation.ts:234]: Fixed in 5db6c25 — the `oneOf` SingleSelect now sets `allowFreeformInput: false`.

File: src/vs/platform/agentHost/node/claude/claudeElicitation.ts [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+					options: field.oneOf.map((o): ChatInputOption => ({ id: o.const, label: o.title || o.const })),
+				};
+			}
+			if (field.enum) {
+				const names = field.enumNames;
+				return {
+					...base,
+					kind: ChatInputQuestionKind.SingleSelect,
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply src/vs/platform/agentHost/node/claude/claudeElicitation.ts:243]: Fixed in 5db6c25 — the plain-`enum` SingleSelect now sets `allowFreeformInput: false`.

File: src/vs/platform/agentHost/node/claude/claudeElicitation.ts [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+		return { action: 'accept' };
+	}
+	const content: Record<string, ElicitationFieldValue> = {};
+	for (const [name] of schema.fields) {
+		const value = elicitationAnswerToValue(answers?.[name]);
+		if (value !== undefined) {
+			content[name] = value;
+		}
+	}
+	return { action: 'accept', content };
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply src/vs/platform/agentHost/node/claude/claudeElicitation.ts]: Good catch — fixed in 5db6c25. Answers are now read with `Object.hasOwn(answers, name)` (so an untrusted `__proto__`/`constructor` name never reads an inherited member and crashes), and `content` is built via `Object.fromEntries` (define semantics, so such a name lands as an own data property rather than mutating the prototype). New test: `elicitationResultFromAnswers is safe against prototype-polluting field names` (uses `JSON.parse` to create real own `__proto__`/`constructor` keys).

File: src/vs/platform/agentHost/node/claude/claudeElicitation.ts [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+ *   {@link ChatInputQuestion}. A missing/malformed schema falls back to a
+ *   message-only request so the user can still accept or decline.
+ */
+export function buildElicitationRequest(requestId: string, request: ElicitationRequest): ChatInputRequest {
+	if (request.mode === 'url') {
+		const result: ChatInputRequest = { id: requestId, message: request.message };
+		if (request.url) {
+			result.url = request.url;
+		}
+		return result;
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply src/vs/platform/agentHost/node/claude/claudeElicitation.ts:142]: Fixed in 5db6c25 — the bridge now cancels a request that has neither a URL nor questions (a `url`-mode request missing its URL, or a form with no representable fields) instead of surfacing the consumer's injected required text question and falsely reporting `accept`. New agent test: `a url-mode request with no url cancels without surfacing a prompt`.

File: src/vs/platform/agentHost/node/claude/claudeElicitation.ts [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+	}
+	const schema = parseElicitationSchema(request.requestedSchema);
+	if (!schema || schema.fields.length === 0) {
+		return { id: requestId, message: request.message };
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply src/vs/platform/agentHost/node/claude/claudeElicitation.ts:146]: Fixed in 5db6c25 together with the URL case — same root cause (the consumer injects a required `answer` text question for question-less requests, which the mapper then discards). The bridge now cancels a request with neither a URL nor questions rather than forcing meaningless text. New agent test: `a form with no representable fields cancels without surfacing a prompt`.

File: src/vs/platform/agentHost/node/claude/roadmap.md [hunk 8]
```diff
@@ -1285,6 +1285,77 @@ dispose) clean across the whole session lifecycle.
...
+- Add a pure-projection module (e.g. `claudeElicitation.ts`, sibling to
+  `claudeInteractiveTools.ts`) for schema→questions and answers→content, so it
+  is unit-testable without standing up an agent.
+- Observe the SDK's per-request `signal` → `respondToUserInputRequest(id,
+  Cancel)` (mirror the abort wiring in
+  [claudeCanUseTool.ts:88](./claudeCanUseTool.ts)); keep the auto-cancel only
+  when unattended (headless / autopilot).
+- URL-mode elicitations complete out-of-band via `SDKElicitationCompleteMessage`
+  ([CONTEXT.md:1096](./CONTEXT.md)) — route it through the message router so the
+  renderer's "open URL" affordance resolves.
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply src/vs/platform/agentHost/node/claude/roadmap.md]: Fixed in 5db6c25 — reworded the roadmap bullet to match the implemented scope: the `onElicitation` promise is awaited so URL prompts resolve via the existing `ChatInputCompleted` path, and `SDKElicitationCompleteMessage` router work is explicitly out of scope (consistent with phase10.6-plan.md).

File: src/vs/platform/agentHost/node/claude/phase10.6-plan.md [hunk 9]
```diff
@@ -0,0 +1,422 @@
...
+- **`IElicitationField` is derived from a base-layer runtime validator and pinned to the MCP SDK's `PrimitiveSchemaDefinition`.** The final design uses `vs/base/common/validation.ts` (the pure "zod-lite") to build `vElicitationField`, runtime-validates each schema field (dropping malformed ones — e.g. a non-array `enum` that would otherwise reach `.map` and throw), and *derives* the field type via `ValidatorType<typeof vElicitationField>` (no hand-rolled interface). A **type-only** import of `PrimitiveSchemaDefinition` plus the compile-time guard `_assertElicitationFieldCoversSchema` (verified non-vacuous and drift-catching) pins the derived type to the SDK's authoritative union. The base validator is used instead of the SDK's own zod schema because a *runtime* `@modelcontextprotocol/sdk` import doesn't resolve in the unit-test renderer (all SDK runtime access goes through `IClaudeAgentSdkService`); a `safeParse` prototype was rejected for exactly that reason. Net vs. the original hand-rolled interface: real runtime validation (fixes a latent throw), zero duplicated type, and spec-drift protection.
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply src/vs/platform/agentHost/node/claude/phase10.6-plan.md:422]: Fixed in 5db6c25 — reconciled the plan's Implementation Notes with the final suite (removed the stale "12 tests / six-test bridge suite" and "no-session untested" claims; the no-session guard is now covered).

File: src/vs/platform/agentHost/node/claude/claudeElicitation.ts [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+
+/**
+ * Compile-time guard: every member of the MCP SDK's authoritative
+ * {@link PrimitiveSchemaDefinition} union must be assignable to
+ * {@link IElicitationField}. If the SDK adds or reshapes a field, this function
+ * fails to compile — forcing {@link vElicitationField} (and the projection
+ * below) to keep up rather than silently under-reading a new schema shape. It is
+ * type-only: never called, and erased at runtime.
+ */
+function _assertElicitationFieldCoversSchema(field: PrimitiveSchemaDefinition): IElicitationField {
```
[the remark this answers is not in the document]
Comment (TylerLeonhardt) [reply src/vs/platform/agentHost/node/claude/claudeElicitation.ts:86]: You're right — corrected in 5db6c25. The docstring no longer claims it catches additions. It now states accurately that the guard catches an *incompatible reshape* of a field we already project (e.g. the SDK retyping `enum` from `string[]` to `number[]`), but NOT a purely additive new union member/keyword, since that stays structurally assignable to the all-optional view and the validator's optional `type: string` accepts any discriminator. The module-level "cannot silently drift" wording was softened to match.

**Review by benvillalobos — APPROVED** [review]
